Add clickable token address

cowdan opened the pull request:
# Summary

Add to a token contract address a link (at the moment only Etherscan is supported) and a `copy to clipboard` icon


<img width="477" alt="Screenshot 2025-02-27 at 15 15 36" src="https://github.com/user-attachments/assets/f4b602f9-ec2d-457b-b997-783eccdaf434" />

<img width="466" alt="Screenshot 2025-02-27 at 15 16 07" src="https://github.com/user-attachments/assets/3b7d4bd7-d55b-4fd3-8a85-9010a05b6313" />

<img width="468" alt="Screenshot 2025-02-27 at 15 16 17" src="https://github.com/user-attachments/assets/5cfd016f-6850-4f85-8e49-fb5516922f0e" />

# To Test

1.  Go to the token list selector

- [ ] Next to the token ticker, there is a shortened contract address. The `copy` icon should not be displayed
- [ ] On mouse over the address, it should be darkened and underlined. Also, the `copy` icon should be displayed.
- [ ] On address link click, it should open the contract on Etherscan in a new tab. The token list should still be there (no going back to the main swap screen). Also, the selected token shouldn't change.
- [ ] On `copy` icon click, it should display, after very short delay, the text "Copied!" and then going back to show the `copy` icon. The token address should be available on the clipboard memory.

# Background

The `disabled` attribute on `TokenListItem` stopped propagating a click event generated by the `CopyToClipboard` component, so I had to remove it and replicate the behaviour (styling and click propagation)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced an interactive address display that shows a shortened blockchain address with tooltip details and copy-to-clipboard functionality.
  - Integrated this clickable address into token information views for quick access to blockchain explorer links.

- **Enhancements**
  - Improved token selection behavior with refined click handling for a smoother user experience.

- **Style**
  - Added new styled components for better layout and hover effects, enhancing visual cues for interactive and selected states.

- **Documentation**
  - Minor punctuation corrections made to enhance clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

cowdan commented:
recheck

cowdan commented:
I have read the CLA Document and I hereby sign the CLA

recheck

anxolin commented on README.md:
```diff
@@ -289,7 +289,7 @@ In case of problems with the service worker cache you force a reset using
 
 Since this repo includes multiple apps, we do not want to build all of them on each PR because it causes long build queues in Vercel.  
-Some apps (see the list bellow) are not required to be built on each PR so we run them only a PR is labeled with a specific label.  
+Some apps (see the list bellow) are not required to be built on each PR so we run them only a PR is labeled with a specific label..
```
The space we had was not needed, but you added another dot?

cowdan replied:
fixed

anxolin commented on apps/cowswap-frontend/src/common/pure/CopyToClipboard/styled.ts:
```diff
@@ -0,0 +1,34 @@
+  margin-left: 4px;
+  max-width: 20px;
+  background-color: transparent;
+  background-image: url('images/click-to-copy.svg');
```
Didn't we have already a copy icon https://github.com/cowprotocol/cowswap/blob/main/apps/cowswap-frontend/src/modules/account/containers/CopyHelper/index.tsx#L10

fairlighteth replied:
Yes, we use `import { Copy } from 'react-feather'` on this `cowswap-frontend` app specifically.

cowdan replied:
Fixed, I've used the `CopyHelper`

anxolin reviewed (commented):
Great to see your first contribution. 

I added a few comments.


I think the color of the button is not right for dark mode:
<img width="512" alt="image" src="https://github.com/user-attachments/assets/56c26db2-4196-4a11-8e7a-b1261ee11ebc" />

Also, are we using this tooltip?
<img width="679" alt="image" src="https://github.com/user-attachments/assets/89cda31a-6f9f-4e88-b7b0-e2c69dd31dec" />

Wouldn't we want to use our normal tooltips?
<img width="750" alt="image" src="https://github.com/user-attachments/assets/e391b2eb-840c-4faa-8957-999745d9948a" />

fairlighteth commented on apps/cowswap-frontend/src/common/pure/ClickableAddress/styled.ts:
```diff
@@ -0,0 +1,29 @@
+  align-items: center;
+
+  &:hover {
+    .clickable-address-copy-to-clipboard {
```
hard coded class name required?

cowdan replied:
right, removed

fairlighteth commented:
- I think adding a new click to copy is probably redundant here? We already have `apps/cowswap-frontend/src/legacy/components/Copy/CopyMod.tsx` in this app. E.g used in `apps/cowswap-frontend/src/pages/Account/Balances.tsx` L257.

I'd suggest to use that and/or take this as an opportunity to refactor (move away from 'legacy' and/or make it somehow a shared lib if appropriate).

- IMO for the native token (e.g. ETH) better to omit the token address? Right now on click it links to something not related to the native token https://etherscan.io/address/0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE 
<img width="202" alt="Screenshot 2025-02-27 at 17 52 24" src="https://github.com/user-attachments/assets/0ffc5b4a-c56a-44b7-954a-283541404017" />

- Darkmode needs the same text color as in light mode <img width="122" alt="Screenshot 2025-02-27 at 17 59 50" src="https://github.com/user-attachments/assets/e1fcfb91-bb6e-45fc-827f-229eadfb7fc2" />

- Hover background color is gone now. This PR:
<img width="506" alt="Screenshot 2025-02-27 at 18 01 27" src="https://github.com/user-attachments/assets/6072577c-68e8-40aa-ad09-0e79e7a49948" />

Develop: 

<img width="641" alt="Screenshot 2025-02-27 at 18 01 59" src="https://github.com/user-attachments/assets/dcae062f-f5ad-446e-94e6-b4f04bca3812" />

shoom3301 commented on apps/cowswap-frontend/src/common/pure/ClickableAddress/index.tsx:
```diff
@@ -0,0 +1,56 @@
+}
+
+// TODO: once we have more chains, we should use the chainId to get the correct explorer
+const mapChainIdToExplorer: { [key: string]: { name: string; url: string } } = {
```
p.s. for such type of maps we use `Record<SupportedChainId, ValueType>` type in order to be safe with adding new networks. With this type If we add a new network we will see compilation errors for all necessary maps.

cowdan replied:
Fixed, used `getChainInfo` method

shoom3301 commented on apps/cowswap-frontend/src/common/pure/ClickableAddress/index.tsx:
```diff
@@ -0,0 +1,56 @@
+
+  const shortAddress = shortenAddress(address)
+
+  const target = useMemo(() => getTarget(address, chainId), [address, chainId])
```
useMemo is not necessary here

cowdan replied:
removed

elena-zh reviewed (changes requested):
Hey @cowdan , great job. thank you!!
However, some issues from my side: 
1. When I open a token list in another chain, the app is crashed 
![crash](https://github.com/user-attachments/assets/afb3e107-95bd-4554-80a9-33c61a96b0d8)
To reproduce: 
- No need to connect to a wallet
- Change a chain using network selector
- Open a token list in the sell/buy field

2. WDYT about changing a chain name to a more generic word like 'block explorer'? If we decide to show addresses in other chains, there will no need to change this word for every chain
![image](https://github.com/user-attachments/assets/6d48eb99-2bea-4c81-9049-aeda0f68f4d3)
 3. In the dark mode 'copy' icon is barely visible: 
![image](https://github.com/user-attachments/assets/395d2226-e554-406f-871a-52b007bfa263)
4. It is impossible to copy an address in a mobile view: when I press on the coly button (on a real device), a token is picked. @fairlighteth , should we still have a copy button in a mobile view? 
![image](https://github.com/user-attachments/assets/7886cac5-6705-409f-8ee8-bac9f71a91f0)


Thanks!

cowdan commented:
> Great to see your first contribution.
> 
> I added a few comments.
> 
> I think the color of the button is not right for dark mode:`

Fixed

> Also, are we using this tooltip? 

It was the anchor element `title` attribute

> Wouldn't we want to use our normal tooltips? 
Right, done

cowdan commented:
> * I think adding a new click to copy is probably redundant here? We already have `apps/cowswap-frontend/src/legacy/components/Copy/CopyMod.tsx` in this app. E.g used in `apps/cowswap-frontend/src/pages/Account/Balances.tsx` L257.
> 
> I'd suggest to use that and/or take this as an opportunity to refactor (move away from 'legacy' and/or make it somehow a shared lib if appropriate).

Right, I've used that component. At this time refactoring it would take longer time, I don't think it's critical at this time

> * IMO for the native token (e.g. ETH) better to omit the token address? Right now on click it links to something not related to the native token https://etherscan.io/address/0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE
> 

Done

> * Darkmode needs the same text color as in light mode 

Done

> * Hover background color is gone now. This PR:

Fixed

cowdan commented:
> Hey @cowdan , great job. thank you!! However, some issues from my side:
> 
> 1. When I open a token list in another chain, the app is crashed

This is fixed now, it was caused by me not using CHAIN-INFO

> 
> 2. WDYT about changing a chain name to a more generic word like 'block explorer'? If we decide to show addresses in other chains, there will no need to change this word for every chain

In CHAIN-INFO we have the explorer names, so it should be ok now

> 3. In the dark mode 'copy' icon is barely visible:
Fixed

> 4. It is impossible to copy an address in a mobile view: when I press on the coly button (on a real device), a token is picked. @fairlighteth , should we still have a copy button in a mobile view?

Not sure how to address this in our current implementation. Maybe a specific fix for mobile?

cowdan commented:
I have read the CLA Document and I hereby sign the CLA

recheck

cowdan commented:
I have read the CLA Document and I hereby sign the CLA

recheck

cowdan commented:
recheck

cowdan commented:
I have read the CLA Document and I hereby sign the CLA

recheck

shoom3301 commented on apps/cowswap-frontend/src/common/pure/ClickableAddress/index.tsx:
```diff
@@ -0,0 +1,59 @@
+
+import * as styledEl from './styled'
+
+const getTarget = (address: string, chainId: number) => {
```
Nitpick, we decided to simplify the way we work with chainId in the app.
We always assume that chainId is `SupportedChainId`, so no need to check it every time.
We do the checks in the top level using `useIsProviderNetworkUnsupported()` so we don't need to check it on every single case.

cowdan commented:
I have read the CLA Document and I hereby sign the CLA

elena-zh reviewed (commented):
Hey @cowdan , great job!!!
I have only 2 tiny nitpicks to report: 
1. WDYT about adding 'Sepolia' to the Etherscan in the tooltip when a user is connected to this network? 
![sepolia Etherscan](https://github.com/user-attachments/assets/49ed8ab3-6f62-4e98-92ed-2f7282cfaee7)
2. Sometime I face difficulties to press on the copy button even in a desktop view: instead of copying, the token is picked into the field. See the [video ](https://www.loom.com/share/12184d893bc942d686517f410150d417). Might it be possible to fix this somehow? E.g. increase the 'copy' tap area or so? 

Thank you!

cowdan commented:
> Hey @cowdan , great job!!! I have only 2 tiny nitpicks to report:
> 
> 1. WDYT about adding 'Sepolia' to the Etherscan in the tooltip when a user is connected to this network?
Yes, done

> 2. Sometime I face difficulties to press on the copy button even in a desktop view: instead of copying, the token is picked into the field. See the [video ](https://www.loom.com/share/12184d893bc942d686517f410150d417). Might it be possible to fix this somehow? E.g. increase the 'copy' tap area or so?
I now changed the tooltip position to the bottom of the address

elena-zh commented:
Hey @cowdan , the 1st case LGTM, thank you!
WDYT about removing brackets from 'Sepolia' word there? 
![image](https://github.com/user-attachments/assets/ead03403-e9ba-4fc1-aa01-576b8a3c0505)

As for the 2nd issue, I still can face it: each time I click on the copy button, a token is picked into the field. See the video: https://www.loom.com/share/60b6f9e029d741e9b73de07c03e84b32

cowdan commented:
> Hey @cowdan , the 1st case LGTM, thank you! WDYT about removing brackets from 'Sepolia' word there? 

Done

> 
> As for the 2nd issue, I still can face it: each time I click on the copy button, a token is picked into the field. See the video: 

This is fixed now!

elena-zh reviewed (approved):
Great job, thank you!

anxolin commented:
Congratulation on your first PR getting merged
